src/basic/fd-util.c

```diff
@@ -913,6 +913,36 @@ int fd_is_opath(int fd) {
         return FLAGS_SET(r, O_PATH);
 }
 
+int fd_verify_safe_flags(int fd) {
+        int flags, unexpected_flags;
+
+        /* Check if an extrinsic fd is safe to work on (by a privileged service). This ensures that clients
+         * can't trick a privileged service into giving access to a file the client doesn't already have
+         * access to (especially via something like O_PATH).
+         *
+         * O_NOFOLLOW: For some reason the kernel will return this flag from fcntl; it doesn't go away
+         *             immediately after open(). It should have no effect whatsoever to an already-opened FD,
+         *             and since we refuse O_PATH it should be safe.
+         *
+         * RAW_O_LARGEFILE: glibc secretly sets this and neglects to hide it from us if we call fcntl.
+         *                  See comment in missing_fcntl.h for more details about this.
+         */
+
+        assert(fd >= 0);
+
+        flags = fcntl(fd, F_GETFL);
+        if (flags < 0)
+                return -errno;
+
+        unexpected_flags = flags & ~(O_ACCMODE|O_NOFOLLOW|RAW_O_LARGEFILE);
+        if (unexpected_flags != 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(EREMOTEIO),
+                                       "Unexpected flags set for extrinsic fd: 0x%x",
```

Comment:
You should probably use octal?

The kernel defines the flags that can be returned from here in octal: https://elixir.bootlin.com/linux/latest/source/include/uapi/asm-generic/fcntl.h#L19

So it would be quite useful to print in octal. If it says `Unexpected flag set for extrinsic fd: 010000000` you can quite easily cross-reference that with O_PATH (just go to that page, and the arch-specific fcntl.h file too, and Ctrl+F). If instead it says `Unexpected flags set for extrinsic fd: 0x200000` it's a bit harder since you'll need to convert it into octal first.

Reply:
@poettering suggested using hex above. I don't have a preference whatsoever.

Reply:
+1 for octal.